processor/k8sattributesprocessor/internal/kube/client.go

```diff
@@ -391,29 +402,26 @@ func (c *WatchClient) extractPodContainersAttributes(pod *api_v1.Pod) map[string
 			if c.Rules.ContainerImageTag && nameTagSep > 0 {
 				container.ImageTag = spec.Image[nameTagSep+1:]
 			}
-			containers[spec.Name] = container
+			containers.ByName[spec.Name] = container
 		}
 	}
-
-	if c.Rules.ContainerID {
-		for _, apiStatus := range append(pod.Status.ContainerStatuses, pod.Status.InitContainerStatuses...) {
-			container, ok := containers[apiStatus.Name]
-			if !ok {
-				container = &Container{}
-				containers[apiStatus.Name] = container
-			}
+	for _, apiStatus := range append(pod.Status.ContainerStatuses, pod.Status.InitContainerStatuses...) {
+		container, ok := containers.ByName[apiStatus.Name]
+		if !ok {
+			container = newContainer(apiStatus.Name)
+			containers.ByName[apiStatus.Name] = container
+		}
+		containerID := apiStatus.ContainerID
```

Comment:
And this becomes 
```suggestion
		container, ok := containers.ByName[apiStatus.Name]
		if !ok {
			container = &Container{}
			containers.ByName[apiStatus.Name] = container
		}
		if c.Rules.ContainerName {
			container.Name = name
		}
		
		containerID := apiStatus.ContainerID
```

so we don't need `newContainer` func anymore

Reply:
True. This works. It's a bit confusing. That's why I [suggested](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20340#discussion_r1149112774) to remove all the conditions related to rules from this package. I'll try to address that separately.

Thanks for your patience.